Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-217.ca-central-1.compute.internal
Sha: 93045be7fed7b2669eaf3ca0c15bcd1042797027
Number of Error(s): 7
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
/home/centos/smoketest/PR-13574/HPCC-Platform/system/jhtree/ctfile.cpp:109:21: error: \xe2\x80\x98struct KeyHdr\xe2\x80\x99 has no member named \xe2\x80\x98nodesize\xe2\x80\x99; did you mean \xe2\x80\x98nodeSize\xe2\x80\x99?
make[2]: *** [system/jhtree/CMakeFiles/jhtree.dir/build.make:76: system/jhtree/CMakeFiles/jhtree.dir/ctfile.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1127: system/jhtree/CMakeFiles/jhtree.dir/all] Error 2
make: *** [Makefile:163: all] Error 2

[Build error:](https://gist.github.com/6d9ee4d7a8e1138929f7010a1f4f4575)
Build log [Build log](https://gist.github.com/HPCCSmoketest/6d9ee4d7a8e1138929f7010a1f4f4575#file-result-20-03-31-17-51-00-log)

--- a/esp/src/eclwatch/FilterDropDownWidget.js
+++ b/esp/src/eclwatch/FilterDropDownWidget.js
@@ -69,6 +69,8 @@ define([
         },
 
         _onFilterApply: function (event) {
+            var formData = this.toObject(event.currentTarget.form);
+            this.initRecentFilter(formData);
             this.filterDropDown.closeDropDown();
             this.emit("apply");
             this.refreshState();
@@ -105,13 +107,13 @@ define([
             return false;
         },
 
-        toObject: function () {
+        toObject: function (filter) {
             var context = this;
             if (this.filterDropDown.get("disabled")) {
                 return {};
             }
             var retVal = {};
-            arrayUtil.forEach(this.filterForm.getDescendants(), function (item, idx) {
+            arrayUtil.forEach(this.filterForm.getDescendants() || filter, function (item, idx) {
                 var name = item.get("name");
                 if (name) {
                     var value = item.get("value");
@@ -120,15 +122,6 @@ define([
                     }
                 }
             });
-            if (this.userName !== null) {
-                if (!Utility.isObjectEmpty(retVal)) {
-                    Recent.addToStack(this.params.ws_key, retVal, 5).then(function (val) {
-                        if (val) {
-                            context.loadRecentFilters(retVal);
-                        }
-                    });
-                }
-            }
             return retVal;
         },
 
@@ -150,11 +143,13 @@ define([
                 return;
 
             this.userName = dojoConfig.username;
+            var recentFilterLoaded = false;
             if (this.userName !== null) {
                 this.recentFilterNode = dom.byId(this.id + "RecentFilters");
 
-                if (params.widget) {
+                if (params.widget && !recentFilterLoaded) {
                     srcReact.render(srcReact.RecentFilters, { ws_key: params.ws_key, widget: params.widget, filter: {} }, this.recentFilterNode);
+                    recentFilterLoaded = true;
                 }
             }
         },
@@ -174,6 +169,17 @@ define([
 
         reset: function (disable) {
             this.filterForm.reset();
+        },
+
+        initRecentFilter: function (retVal) {
+            var context = this;
+            if (this.userName !== null) {
+                if (!Utility.isObjectEmpty(retVal)) {
+                    Recent.addToStack(this.params.ws_key, retVal, 5, true).then(function(val){
+                        context.loadRecentFilters(retVal);
+                    });
+                }
+            }
         },
 
         loadRecentFilters: function (retVal) {

--- a/esp/src/src/UserPreferences/Recent.ts
+++ b/esp/src/src/UserPreferences/Recent.ts
@@ -1,8 +1,9 @@
 import { userKeyValStore } from "../KeyValStore";
+import { hashSum } from "@hpcc-js/util";
 
 const ws_store = userKeyValStore();
 
-export function addToStack(key: string, data: any, expectedLength?: number) {
+export function addToStack(key: string, data: any, expectedLength?: number, removeDuplicates?: boolean) {
     let finalData;
     return new Promise(function(resolve, reject) {
         ws_store.get(key)
@@ -12,15 +13,17 @@ export function addToStack(key: string, data: any, expectedLength?: number) {
                 resolve(finalData = JSON.stringify([data]));
             } else {
                 const encodedData = JSON.parse(response);
-                if (encodedData?.length >= expectedLength) {
-                    encodedData.pop();
-                    encodedData.unshift(data);
-                    ws_store.set(key, JSON.stringify(encodedData));
-                    resolve(finalData = JSON.stringify([data]));
-                } else if (encodedData.length >= 1) {
-                    encodedData.unshift(data);
-                    ws_store.set(key,  JSON.stringify(encodedData));
-                    resolve(finalData = JSON.stringify([data]));
+                if (removeDuplicates && hashSum(encodedData[0]) !== hashSum(data)) {
+                    if (encodedData?.length >= expectedLength) {
+                        encodedData.pop();
+                        encodedData.unshift(data);
+                        ws_store.set(key, JSON.stringify(encodedData));
+                        resolve(finalData = JSON.stringify([data]));
+                    } else if (encodedData.length >= 1) {
+                        encodedData.unshift(data);
+                        ws_store.set(key,  JSON.stringify(encodedData));
+                        resolve(finalData = JSON.stringify([data]));
+                    }
                 }
             }
         });

